### PR TITLE
[SMALLFIX] Refactored InodeTree.traverseToInode

### PR DESCRIPTION
**Pull request by yupeng9**

### CODE REVIEW

**Comment by calvinjia** [comment]
@yupeng9 Could you please tag this with a JIRA ticket or `SMALLFIX` if its a simple change?

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java [hunk 1]
```diff
@@ -253,57 +251,87 @@ public void setPinned(Inode inode, boolean pinned) {
...
+      mInode = inode;
+    }
+
+    boolean isFound() {
+      return mFound;
+    }
+
+    int getNonexistentPathIndex() {
+      if (mFound) {
+        throw new UnsupportedOperationException("The trasveral is successful");
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: "trasveral" -> "traversal"

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java [hunk 2]
```diff
@@ -253,57 +251,87 @@ public void setPinned(Inode inode, boolean pinned) {
...
       }
     }
-    return ret;
+    return TraverseResult.createFoundResult(current);
   }
 
+  private static final class TraverseResult {
+    private final boolean mFound;
+    /**
+     * when the path is not found in a traversal, the index of the last navigated step in the path
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: Is this index the first one NOT found, or the last one found? I think the comments are not clear.

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java [hunk 3]
```diff
@@ -253,57 +251,87 @@ public void setPinned(Inode inode, boolean pinned) {
...
+      } else {
+        // next is a directory and keep navigating
+        current = next;
       }
     }
-    return ret;
+    return TraverseResult.createFoundResult(current);
   }
 
+  private static final class TraverseResult {
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: Should this be "TraversalResult"?

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java [hunk 4]
```diff
@@ -253,57 +251,87 @@ public void setPinned(Inode inode, boolean pinned) {
...
-    for (int k = 1; k < pathComponents.length; k ++) {
-      Inode next = ((InodeDirectory) ret.getFirst()).getChild(pathComponents[k]);
+    for (int i = 1; i < pathComponents.length; i ++) {
+      Inode next = ((InodeDirectory) current).getChild(pathComponents[i]);
       if (next == null) {
-        // The user might want to create the nonexistent directories, so we leave ret.getFirst()
-        // as the last Inode taken. We set nonexistentInd to k, to indicate that the kth path
+        // The user might want to create the nonexistent directories, so return the current inode
+        // directory
+        // as the last Inode taken. Also set nonexistentInd to k, to indicate that the kth path
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: Please modify the comments to match the new code ("nonexistentInd", "k").

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java [hunk 5]
```diff
@@ -253,57 +251,87 @@ public void setPinned(Inode inode, boolean pinned) {
...
+        throw new InvalidPathException("File name starts with " + pathComponents[0]);
       }
     }
 
-    Pair<Inode, Integer> ret = new Pair<Inode, Integer>(mRoot, -1);
+    Inode current = mRoot;
 
-    for (int k = 1; k < pathComponents.length; k ++) {
-      Inode next = ((InodeDirectory) ret.getFirst()).getChild(pathComponents[k]);
+    for (int i = 1; i < pathComponents.length; i ++) {
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: A comment why it starts from "1" and not "0"?

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java [hunk 6]
```diff
@@ -123,30 +121,30 @@ public Inode createPath(TachyonURI path, long blockSizeBytes, boolean recursive,
...
-        final String msg =
-            "File " + path + " creation failed. Component " + inodeTraversal.getSecond() + "("
-                + parentPath[inodeTraversal.getSecond()] + ") does not exist";
+        final String msg = "File " + path + " creation failed. Component "
+            + traversalResult.getNonexistentPathIndex() + "("
+            + parentPath[traversalResult.getNonexistentPathIndex()] + ") does not exist";
         LOG.info("InvalidPathException: " + msg);
         throw new InvalidPathException(msg);
       } else {
         // We will start filling in the path from inodeTraversal.getSecond()
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: Please update this comment.

**Comment by gpang** [comment]
Thanks for this improvement @yupeng9 ! I left a few comments.

**Comment by yupeng9** [comment]
Thanks @gpang for the comments. I have addressed them all.

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java [hunk 7]
```diff
@@ -253,57 +252,87 @@ public void setPinned(Inode inode, boolean pinned) {
...
+      mInode = inode;
+    }
+
+    boolean isFound() {
+      return mFound;
+    }
+
+    int getNonexistentPathIndex() {
+      if (mFound) {
+        throw new UnsupportedOperationException("The trasversal is successful");
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: "trasversal" -> "traversal"

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeDirectoryTests.java [hunk 8]
```diff
@@ -104,9 +104,11 @@ public void removeNonExistentChildTest() {
   @Test
   public void deleteInodeTest() {
     InodeDirectory inode1 = createInodeDirectory();
-    Assert.assertEquals(false, inode1.isDeleted());
+    Assert.assertFalse(inode1.isDeleted());
     inode1.delete();
-    Assert.assertEquals(true, inode1.isDeleted());
+    Assert.assertTrue(inode1.isDeleted());
+    inode1.restore();
```
Comment (gpang) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeDirectoryTests.java]: Was restore() added? The changes in this file seem unrelated to the traversal change.

**Comment by gpang** [comment]
I have a few additional comments.

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java (hunk 7 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: done

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeDirectoryTests.java (hunk 8 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeDirectoryTests.java]: This is because when I checked out the branch, the branch was in an inconsistent state. And the changes in this file were to resolve inconsistency.

**Comment by gpang** [comment]
LGTM

**Comment by calvinjia** [comment]
+1 LGTM
